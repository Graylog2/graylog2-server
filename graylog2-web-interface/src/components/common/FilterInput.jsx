import PropTypes from 'prop-types';
import React from 'react';

import { Input } from 'components/bootstrap';

/**
 * The FilterInput component will render a filter input and takes a
 * list or object to be filtered and a filter function which will be
 * executed while typing.
 */
class FilterInput extends React.Component {
  static propTypes = {
    /**
     * This callback will be called when the user types in a filter string.
     * It takes the new filter string as an argument.
     */
    onChange: PropTypes.func.isRequired,
    /** The optional label for the filter. Default is 'Filter' */
    filterLabel: PropTypes.string,
    /** The wrapperClassName of the Input for styling */
    wrapperClassName: PropTypes.string,
    /** The labelClassName of the Input for styling */
    labelClassName: PropTypes.string,
  };

  static defaultProps = {
    filterLabel: 'Filter',
    wrapperClassName: 'col-sm-4',
    labelClassName: 'col-sm-1',
  };

  constructor(props) {
    super(props);

    this.state = {
      filter: '',
    };
  }

  _onChange = (e) => {
    const filter = e.target.value;
    this.setState({ filter: filter });
    this.props.onChange(filter);
  };

  render() {
    return (
      <form className="form-horizontal" onSubmit={(e) => { e.preventDefault(); }}>
        <Input
          id="filter-descriptors"
          type="text"
          wrapperClassName={this.props.wrapperClassName}
          labelClassName={this.props.labelClassName}
          label={this.props.filterLabel}
          value={this.state.filter}
          onChange={this._onChange}
        />
      </form>
    );
  }
}

export default FilterInput;
