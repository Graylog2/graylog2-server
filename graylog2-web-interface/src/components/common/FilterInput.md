```js
const createReactClass = require('create-react-class');

const FilterInputExample = createReactClass({

  getInitialState() {
    return {
      filtered_rows: this.rows(),
    };
  },

  rows() {
    return [
      "Rand",
      "Mat",
      "Perrin",
      "Egwene",
      "Nynaeve",
    ]
  },

  onChange(filter) {
    if (filter === '') {
      this.setState({ filtered_rows: this.rows() });
      return;
    }
    const newRows = this.state.filtered_rows.filter((row) => {
      const regexp = RegExp(filter, 'i');
      return regexp.test(row);
    });
    this.setState({ filtered_rows: newRows });
  },

  render() {
    const rows = this.state.filtered_rows.map(e => <li key={e}>{e}</li>);
    return (
      <div>
        <FilterInput onChange={this.onChange} />
        <ul>
          {rows}
        </ul>
      </div>
    );
  },
});

<FilterInputExample />
```
