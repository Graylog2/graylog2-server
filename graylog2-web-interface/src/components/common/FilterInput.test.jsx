import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import 'helpers/mocking/react-dom_mock';

import FilterInput from 'components/common/FilterInput';

describe('<FilterInput />', () => {
  it('should render a filter input', () => {
    const wrapper = renderer.create(<FilterInput onChange={() => {}} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render a filter input with custom label', () => {
    const wrapper = renderer.create(<FilterInput
      onChange={() => {}}
      filterLabel="Search"
      labelClassName="col-sm-3"
      wrapperClassName="col-xs-1"
    />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should take filter input', () => {
    const changeFn = jest.fn((value) => {
      expect(value).toEqual('kaladin');
    });
    const wrapper = mount(<FilterInput onChange={changeFn} />);

    wrapper.find('input').simulate('change', { target: { value: 'kaladin' } });
    expect(changeFn.mock.calls.length).toBe(1);
  });
});
