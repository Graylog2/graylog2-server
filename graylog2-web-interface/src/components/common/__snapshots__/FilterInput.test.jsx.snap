// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterInput /> should render a filter input 1`] = `
<form
  className="form-horizontal"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className="col-sm-1 control-label"
      htmlFor="filter-descriptors"
    >
      Filter
    </label>
    <div
      className="col-sm-4"
    >
      <input
        className="form-control"
        id="filter-descriptors"
        label="Filter"
        onChange={[Function]}
        placeholder=""
        type="text"
        value=""
      />
      
    </div>
  </div>
</form>
`;

exports[`<FilterInput /> should render a filter input with custom label 1`] = `
<form
  className="form-horizontal"
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className="col-sm-3 control-label"
      htmlFor="filter-descriptors"
    >
      Search
    </label>
    <div
      className="col-xs-1"
    >
      <input
        className="form-control"
        id="filter-descriptors"
        label="Search"
        onChange={[Function]}
        placeholder=""
        type="text"
        value=""
      />
      
    </div>
  </div>
</form>
`;
